### PR TITLE
Feature/stageclasses

### PR DESCRIPTION
**Pull request by davibicudo**
This PR adresses issues #52 and #53. It doesn't cover the issues entirely however.
@sebhoerl let me know what you think, perhaps you have some other ideas in mind on how to solve the same issues.

### CODE REVIEW

**Comment by davibicudo** [comment]
I was also able to use this when running arbitrary code in Jupyter. Caching worked well and devalidation as well, when the code changed (inspect.getsource seems to habe done it's job). Using "global" make it no longer robust, but may be helpful for interactive tests and caching (although IPython also has other solutions for that): 
image

File: src/synpp/pipeline.py [hunk 1]
```diff
@@ -481,15 +480,16 @@ def update_json(meta, working_directory):
...
     # 0) Construct pipeline config
     pipeline_config = {}
     if "processes" in config: pipeline_config["processes"] = config["processes"]
     if "progress_interval" in config: pipeline_config["progress_interval"] = config["progress_interval"]
 
     if not working_directory is None:
         if not os.path.isdir(working_directory):
-            raise PipelineError("Working directory does not exist: %s" % working_directory)
+            logger.warning("Working directory does not exist, it will be created: %s" % working_directory)
+            os.mkdir(working_directory)
```
Comment (sebhoerl) [inline src/synpp/pipeline.py:495]: I see the point, but so far I was deliberately avoiding this, to make sure that the user is 100% aware where all the caching files will go. Maybe we can add a config flag, which is usually "false" like "auto_generate_working_directory" ?

File: src/synpp/pipeline.py [hunk 2]
```diff
@@ -580,9 +580,11 @@ def run(definitions, config = {}, working_directory = None, flowchart_path = Non
 
     # 4) Devalidate stages
     sorted_cached_hashes = sorted_hashes - ephemeral_counts.keys()
+    stale_hashes = set()
 
-    # 4.1) Devalidate if they are required
-    stale_hashes = set(required_hashes)
+    # 4.1) Devalidate if they are required (optional, otherwise will reload from cache)
+    if rerun_required:
```
Comment (sebhoerl) [inline src/synpp/pipeline.py:589]: Is this covered by a unit test ? If not, can you add one ?

File: src/synpp/pipeline.py [hunk 3]
```diff
@@ -778,26 +788,60 @@ def run(definitions, config = {}, working_directory = None, flowchart_path = Non
...
+        if definitions is None and self.definitions is None:
+            raise PipelineError("A list of stage definitions must be available in object or provided explicitly.")
+        elif definitions is None:
+            definitions = self.definitions
+        if dryrun is None:
+            dryrun = self.dryrun
+        return run(definitions, self.config, self.working_directory, flowchart_path=flowchart_path,
+                   dryrun=dryrun, verbose=verbose, logger=self.logger, rerun_required=rerun_required)
+
+    def run_single(self, descriptor, config={}, rerun_if_cached=False, dryrun=False, verbose=False):
```
Comment (sebhoerl) [inline src/synpp/pipeline.py:820]: Really nice! If it had a unit test, would be even better :)

**Review by sebhoerl — CHANGES_REQUESTED** [review]

**Comment by sebhoerl** [comment]
Really nice, I like the jupyter integration! It would be nice if you could add some unit tests so this is robust for the future. Thanks!

**Comment by davibicudo** [comment]
Sounds good, I'll increment the PR with a couple tests then.

**Review by sebhoerl — APPROVED** [review]

**Comment by sebhoerl** [comment]
Please update the `CHANGELOG.md` then it should be good to go :) Thanks!

**Pull request merged**